File: Python (Flask) Secundary/app/database/seed.py
from ..extensions import db
from ..modules.brands.models import Brand
from ..modules.models.models import Model
from ..modules.cars.models import Car, Fuel
from ..utils import normalize_upper_no_accent
from datetime import datetime

def epoch_to_dt(ts: int) -> datetime:
    s = int(ts)
    digits = len(str(abs(s)))
    if digits >= 13: return datetime.fromtimestamp(s/1000)
    return datetime.fromtimestamp(s)

def parse_valor(v):
    if isinstance(v,(int,float)): return float(v)
    return float(v.replace(".","").replace(",", "."))

input_data = {
  "cars":[
    {"id":1,"timestamp_cadastro":1696539488,"modelo_id":12,"ano":2015,"combustivel":"FLEX","num_portas":4,"cor":"BEGE","nome_modelo":"ONIX PLUS","valor":"50.000"},
    {"id":2,"timestamp_cadastro":1696531234,"modelo_id":14,"ano":2014,"combustivel":"FLEX","num_portas":4,"cor":"AZUL","nome_modelo":"JETTA","valor":"49.000"},
    {"id":3,"timestamp_cadastro":16965354321,"modelo_id":79,"ano":1993,"combustivel":"DIESEL","num_portas":4,"cor":"AZUL","nome_modelo":"HILUX SW4","valor":"47.500"}
  ]
}

def guess_brand(model_name: str) -> str:
    n = model_name.upper()
    if "ONIX" in n: return "Chevrolet"
    if "JETTA" in n: return "Volkswagen"
    if "HILUX" in n: return "Toyota"
    return "Marca Desconhecida"

def run_seed():
    brand_by_name = {}
    model_by_id = {}

    for c in input_data["cars"]:
        bname = guess_brand(c["nome_modelo"])
        brand = brand_by_name.get(bname) or Brand.query.filter_by(name=bname).first()
        if not brand:
            brand = Brand(name=bname)
            db.session.add(brand); db.session.flush()
            brand_by_name[bname] = brand

        if c["modelo_id"] not in model_by_id:
            model = Model.query.get(c["modelo_id"])
            if not model:
                model = Model(id=c["modelo_id"], nome=c["nome_modelo"], fipeValue=parse_valor(c["valor"]), brand_id=brand.id)
                db.session.add(model); db.session.flush()
            model_by_id[c["modelo_id"]] = model

    for c in input_data["cars"]:
        model = model_by_id[c["modelo_id"]]
        car = Car.query.get(c["id"])
        if not car:
            car = Car(id=c["id"],
                      createdAt=epoch_to_dt(c["timestamp_cadastro"]),
                      model_id=model.id,
                      ano=c["ano"],
                      combustivel=Fuel[normalize_upper_no_accent(c["combustivel"])],
                      num_portas=c["num_portas"],
                      cor=c["cor"])
            db.session.add(car)
    db.session.commit()
    print("Seed concluído com sucesso.")

File: NodeJs (NestJs) Primary/src/database/seed/seed.ts
import dataSource from '../../../ormconfig';
import { Brand } from '../../modules/brands/entities/brand.entity';
import { Model } from '../../modules/models/entities/model.entity';
import { Car, Fuel } from '../../modules/cars/entities/car.entity';
import { input } from './data.seed';

export type InputCar = {
  id: number;
  timestamp_cadastro: number;
  modelo_id: number;
  ano: number;
  combustivel: keyof typeof Fuel;
  num_portas: number;
  cor: string;
  nome_modelo: string;
  valor: string | number;
};

function guessBrandName(modelName: string): string {
  const n = modelName.toUpperCase();

  if (/(SW4|HILUX|GR COROLLA|COROLLA)/.test(n)) return 'Toyota';

  if (/(RS6|AUDI)/.test(n)) return 'Audi';

  if (/(RANGE ROVER|DEFENDER)/.test(n)) return 'Land Rover';

  if (/\bSEAL\b/.test(n)) return 'BYD';

  if (/(911|MACAN|PANAMERA|CAYENNE)/.test(n)) return 'Porsche';

  if (/(^|\s)1500\sREBEL/.test(n)) return 'RAM';

  if (/(MUSTANG|RANGER RAPTOR)/.test(n)) return 'Ford';

  if (/CIVIC/.test(n)) return 'Honda';

  if (/(^|\s)I4\b|M2 COMPETITION|X6 M/.test(n)) return 'BMW';

  return 'Marca Desconhecida';
}

function parseValor(v: string | number): number {
  if (typeof v === 'number') return v;
  const cleaned = v.replace(/\./g, '').replace(',', '.');
  const num = Number(cleaned);
  if (Number.isNaN(num)) throw new Error(`Valor inválido: ${v}`);
  return num;
}

function epochToDate(ts: number): Date {
  const n = Number(ts);
  const len = String(Math.trunc(Math.abs(n))).length;
  if (len >= 13) return new Date(n);
  if (len <= 10) return new Date(n * 1000);
  return new Date(n > 2_000_000_000 ? n : n * 1000);
}

async function run() {
  await dataSource.initialize();

  const brandRepo = dataSource.getRepository(Brand);
  const modelRepo = dataSource.getRepository(Model);
  const carRepo = dataSource.getRepository(Car);

  const brandByName = new Map<string, Brand>();
  const modelById = new Map<number, Model>();

  for (const c of input.cars) {
    const brandName = guessBrandName(c.nome_modelo);
    let brand =
      brandByName.get(brandName) ??
      (await brandRepo.findOne({ where: { name: brandName } }));

    if (!brand) {
      brand = brandRepo.create({ name: brandName });
      await brandRepo.save(brand);
      brandByName.set(brandName, brand);
    }

    if (!modelById.has(c.modelo_id)) {
      let model = await modelRepo.findOne({
        where: { id: Number(c.modelo_id) },
      });

      if (!model) {
        model = modelRepo.create({
          id: c.modelo_id,
          nome: c.nome_modelo,
          fipeValue: parseValor(c.valor),
          brand,
        });
        await modelRepo.save(model);
      }

      modelById.set(c.modelo_id, model);
    }
  }

  for (const c of input.cars) {
    const model = modelById.get(c.modelo_id)!;

    const car = carRepo.create({
      id: c.id,
      createdAt: epochToDate(c.timestamp_cadastro),
      model,
      ano: c.ano,
      combustivel: c.combustivel.toUpperCase() as Fuel,
      num_portas: c.num_portas,
      cor: c.cor,
    });

    await carRepo.save(car);
  }

  console.log('Seed concluído com sucesso.');
  await dataSource.destroy();
}

run().catch((e) => {
  console.error(e);
  process.exit(1);
});
